keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1330 @@
+- (Idle) -> Cordon -> (Complete) -> _Deletion_.
+  - Make a set of nodes unschedulable and then schedulable again.
+  - The complete stage will always be run even without specifying it.
+- (Idle) -> (Cordon) -> Drain -> (Complete) -> _Deletion_.
+  - Make a set of nodes unschedulable, drain them, and then make them schedulable again.
+  - Cordon and Complete stages will always be run, even without specifying them.
+- (Idle) -> Complete -> _Deletion_.
+  -  Make a set of nodes schedulable.
+
+### Deployment and ReplicaSet Controllers
+
+The replica set controller will watch its pods and count the number of pods it observes with a
+`EvacuationRequested` condition. It will then store this count in `.status.ReplicasToEvacuate`
+
+The deployment controller will watch its ReplicaSets and react when it observes positive number of
+pods in `.status.ReplicasToEvacuate`. If the owning object of the targeted pods is a Deployment with
+a positive `.spec.strategy.rollingUpdate.maxSurge` value, the controller will create surge pods by
+scaling up the ReplicaSet. The new pods will not be scheduled on the maintained node because the
+`.spec.unschedulable` field would be set to true on that node. As soon as the surge pods become
+available, the deployment controller will scale down the ReplicaSet. The replica set controller
+will then in turn delete the pods with the `EvacuationRequested` condition.
+
+For completeness, the deployment controller will also track the total number of targeted pods of
+all its ReplicaSets under its `.status.ReplicasToEvacuate`.
+
+If the node maintenance prematurely ends before the surge process has a chance to complete, the
+deployment controller will scale down the ReplicaSet which will then remove the extra pods that
+were created during the surge.
+
+```golang
+type ReplicaSetStatus struct {
+    ...
+    ReplicasToEvacuate int32
+    ...
+}
+```
+
+```golang
+type DeploymentStatus struct {
+    ...
+    ReplicasToEvacuate int32
+    ...
+}
+```
+
+To support providing a response to the drain process that the evacuation has begun. The deployment
+controller will annotate all replica sets that support the evacuation with the annotation
+`deployment.kubernetes.io/evacuation-ready`. For now this will apply to replica sets of deployments
+with `.spec.strategy.rollingUpdate.maxSurge`. When this annotation is present, the replication
+controller will respond to all evacuation requests by setting the `EvacuationInitiated` condition
+to all of its pods with the `EvacuationRequested` condition.
+
+
+### Node Drain Process: Before and After
+
+The following diagrams describe how the node drain process will change in respective to each
+component.
+
+Current state of node drain:
+
+![drain-before](https://raw.githubusercontent.com/kubernetes/website/f2ef324ac22e5d9378f2824af463777182817ca6/static/images/docs/kubectl_drain.svg)
+
+
+Proposed node drain:
+
+![drain-after](node-maintenance.svg)
```

Comment:
Hi, I come across this proposal and it looks interesting!

From this diagram, it shows that `kubectl drain NODE` creates the `NodeMaintenance` object with `.spec.stage=Drain`. What about `idle` stage? Who would create the object in `idle` to signal the start of planning?

Reply:
Today, I think we should discard the whole kubectl drain flow and move directly into the declarative realm.

Reply:
@wangzhen127 The KEP is using the declarative format now and discarding the kubectl drain use case. PTAL